pkg/agent/controller/networkpolicy/node_reconciler_linux.go

```diff
@@ -0,0 +1,714 @@
+	// - For IPv4, iptables rules installed in the chain ANTREA-INGRESS-RULES for ingress rules.
+	// - For IPv6, ip6tables rules installed in the chain ANTREA-INGRESS-RULES for ingress rules.
+	// - For IPv4, iptables rules installed in the chain ANTREA-EGRESS-RULES for egress rules.
+	// - For IPv6, ip6tables rules installed in the chain ANTREA-EGRESS-RULES for egress rules.
+	return r.coreIPTChains[newChainKey(iptChain, isIPv6)]
+}
+
+func groupMembersToIPNets(groups v1beta2.GroupMemberSet, isIPv6 bool) sets.Set[string] {
+	ipNets := sets.New[string]()
+	suffix := "/32"
+	if isIPv6 {
+		suffix = "/128"
+	}
+	for _, member := range groups {
+		for _, ip := range member.IPs {
+			ipAddr := net.IP(ip)
+			if isIPv6 == utilnet.IsIPv6(ipAddr) {
+				ipNets.Insert(ipAddr.String() + suffix)
+			}
+		}
+	}
+	return ipNets
+}
+
+func ipBlocksToIPNets(ipBlocks []v1beta2.IPBlock, isIPv6 bool) []string {
+	var ipNets []string
+	for _, b := range ipBlocks {
+		blockCIDR := ip.IPNetToNetIPNet(&b.CIDR)
+		if isIPv6 != utilnet.IsIPv6CIDR(blockCIDR) {
+			continue
+		}
+		exceptIPNets := make([]*net.IPNet, 0, len(b.Except))
+		for i := range b.Except {
+			c := b.Except[i]
+			except := ip.IPNetToNetIPNet(&c)
+			exceptIPNets = append(exceptIPNets, except)
+		}
+		diffCIDRs, err := ip.DiffFromCIDRs(blockCIDR, exceptIPNets)
+		if err != nil {
+			klog.ErrorS(err, "Error when computing effective CIDRs by removing except IPNets from IPBlock")
+			continue
+		}
+		for _, d := range diffCIDRs {
+			ipNets = append(ipNets, d.String())
+		}
+	}
+	return ipNets
+}
+
+func getIPNetsFromRule(rule *CompletedRule, isIPv6 bool) sets.Set[string] {
+	var set sets.Set[string]
+	if rule.Direction == v1beta2.DirectionIn {
+		set = groupMembersToIPNets(rule.FromAddresses, isIPv6)
+		set.Insert(ipBlocksToIPNets(rule.From.IPBlocks, isIPv6)...)
+	} else {
+		set = groupMembersToIPNets(rule.ToAddresses, isIPv6)
+		set.Insert(ipBlocksToIPNets(rule.To.IPBlocks, isIPv6)...)
+	}
+	// If the set contains "0.0.0.0/0" or "::/0", it means the rule matches any source or destination IP address, just
+	// return a new set only containing "0.0.0.0/0" or "::/0".
+	if isIPv6 && set.Has(ipv6Any) {
+		return sets.New[string](ipv6Any)
+	}
+	if !isIPv6 && set.Has(ipv4Any) {
+		return sets.New[string](ipv4Any)
+	}
+	return set
+}
+
+func getCoreIPTChain(rule *CompletedRule) string {
+	if rule.Direction == v1beta2.DirectionIn {
+		return config.NodeNetworkPolicyIngressRulesChain
+	}
+	return config.NodeNetworkPolicyEgressRulesChain
+}
+
+func buildCoreIPTRule(ipProtocol iptables.Protocol,
+	iptChain string,
+	ipset string,
+	ipNet string,
+	iptRuleTarget string,
+	iptRuleComment string,
+	service *v1beta2.Service,
+	isIngress bool) string {
+	builder := iptables.NewRuleBuilder(iptChain)
+	if isIngress {
+		if ipset != "" {
+			builder = builder.MatchIPSetSrc(ipset)
+		} else if ipNet != "" {
+			builder = builder.MatchCIDRSrc(ipNet)
+		} else {
+			builder = builder.MatchNoSrc(ipProtocol)
```

Comment:
I am trying to refactor the code as you said. This might introduce more `if`s to distinguish such case. I will push another commit later for this change.

Reply:
Then let's keep it in this PR. It should be a rare case anyway.

Reply:
This might not be very complicated. I have added a commit https://github.com/antrea-io/antrea/pull/5658/commits/a6b43db86d44de6ed99cd384d84f86bef05bc2e2 to implement the case.

Reply:
In above implementation, if a rule has no ip address to match, then core iptables rule will not be synced. The service iptables rules will be added or updated if the rule has multiple services.

Reply:
The handling looks good to me